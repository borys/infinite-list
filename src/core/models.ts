export type OfferId = number;

export enum OfferStatus {
  Published = 'published',
  InProgress = 'in_progress',
  WaitingForApproval = 'waiting_for_approval',
  Canceled = 'canceled',
}

export interface Offer {
  id: OfferId;
  title: string;
  description: string;
  img_url: string;
  price: number;
  discount: number;
  rating: number;
  status: OfferStatus;
  created_at: string;
}
