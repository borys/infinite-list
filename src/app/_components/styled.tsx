import styled from "styled-components";
import Image from "next/image";

export const Wrapper = styled.div`
  border: 1px solid black;
  border-radius: 10px;
  overflow: hidden;
`;

export const ScrollableList = styled.ul`
  display: block;
  width: 100%;
  overflow: hidden;
  max-height: 70vh;
  min-height: 70vh;
  overflow: auto;
  padding: 0;
  margin: 0;
`;

export const Item = styled.li`
  display: grid;
  background-color: azure;
  padding: 10px;
  max-height: 85px;
  border: 1px solid #f0f0f0;
  cursor: pointer;
  grid-template-columns: 110px auto 100px;
  grid-template-rows: auto;
  grid-template-areas:
    "photo title price"
    "photo description price";

  &:hover {
    background-color: rgba(0, 0, 0, 0.03);
  }
`;

export const Avatar = styled(Image)`
  grid-area: photo;
  padding: 5px;
  width: 100px;
  height: 75px;
`;

export const Title = styled.h3`
  grid-area: title;
  font-weight: border;
  font-size: 1.2rem;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  margin: 0;
`;

export const Description = styled.p`
  grid-area: description;
  font-size: 1em;
  font-style: italic;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  max-height: 2em;
  max-width: 600px;
  margin: 0;
`;

export const Price = styled.div`
  grid-area: price;
  font-weight: border;
  font-size: 2rem;
  float: right;
  justify-self: center;
  align-self: center;
`;
