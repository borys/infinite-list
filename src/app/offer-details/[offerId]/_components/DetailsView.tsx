import { Offer } from '@/app/_models/Offer';
import React from 'react';

import {
  Description,
  EmptyState,
  Image,
  Price,
  Summary,
  Title,
} from './styled';

export interface DetailsViewProps {
  offer: Offer | null;
  goBack: () => void;
}

export const DetailsView: React.FC<DetailsViewProps> = ({
  offer,
  goBack,
}: DetailsViewProps) => {
  if (!offer) {
    return <EmptyState>Select offer to view</EmptyState>;
  }

  return (
    <div>
      <button onClick={() => goBack()}>Go back</button>
      <Title>{offer.title}</Title>
      <Image src={offer.img_url} alt='offer image' />
      <Description>{offer.description}</Description>
      <Summary>
        <Price>
          Price: {offer.price}
        </Price>
      </Summary>
    </div>
  );
};
