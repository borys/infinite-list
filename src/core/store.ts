import { AppStore } from 'core/store';
import { OfferDetailsState } from 'offerDetails/reducers';
import offerDetailsReducer, {
  initState as offerDetailsInitState,
} from 'offerDetails/reducers';
import { OfferListState } from 'offerList/reducers';
import offerListReducer, {
  initState as offerListInitState,
} from 'offerList/reducers';
import { applyMiddleware, combineReducers, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';

export interface AppStore {
  offerDetails: OfferDetailsState;
  offerList: OfferListState;
}

export const initAppStore = () => {
  const middleware = [thunk];

  const initState: AppStore = {
    offerDetails: offerDetailsInitState,
    offerList: offerListInitState,
  };

  const reducers = combineReducers({
    offerDetails: offerDetailsReducer,
    offerList: offerListReducer,
  });

  let appliedMiddlewares = applyMiddleware(...middleware);
  if (process.env.NODE_ENV !== 'production') {
    appliedMiddlewares = composeWithDevTools(appliedMiddlewares);
  }

  return createStore(reducers, initState, appliedMiddlewares);
};
