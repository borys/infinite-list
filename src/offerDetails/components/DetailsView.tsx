import { Offer } from 'core/models';
import React from 'react';

import {
  Description,
  EmptyState,
  Image,
  Price,
  Summary,
  Title,
} from './styled';

export interface DetailsViewProps {
  offer: Offer | null;
}

export function DetailsView({ offer }: DetailsViewProps) {
  if (!offer) {
    return <EmptyState>Select offer to view</EmptyState>;
  }

  return (
    <div>
      <Title>{offer.title}</Title>
      <Image src={offer.img_url} />
      <Description>{offer.description}</Description>
      <Summary>
        <Price>
          Cena:
          {offer.price}
        </Price>
      </Summary>
    </div>
  );
}
