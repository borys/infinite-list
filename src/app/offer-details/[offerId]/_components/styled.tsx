import styled from 'styled-components';

export const Title = styled.h1`
  font-size: 1.5rem;
  font-weight: bold;
`;

export const Description = styled.span`
  font-size: 1rem;
`;

export const Image = styled.img`
  float: right;
  margin: 5px;
`;

export const EmptyState = styled.div`
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.1);
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const Summary = styled.div`
  margin-top: 10px;
  border-top: 1px solid black;
  clear: both;
  width: 100%;
`;

export const Price = styled.div`
  font-size: 4rem;
`;
