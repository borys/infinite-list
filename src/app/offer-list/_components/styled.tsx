import styled from 'styled-components';

export const ScrollableContainer = styled.div`
  width: 100%;
  max-height: 90vh;
  min-height: 90vh;
  overflow: auto;
`;

export const Item = styled.div`
  max-height: 75px;
  border: 1px solid #f0f0f0;
  cursor: pointer;
  display: grid;
  grid-template-columns: 100px auto 100px;
  grid-template-rows: auto;
  grid-template-areas:
    'image title price'
    'image description price';

  &:hover {
    background-color: rgba(0, 0, 0, 0.1);
  }
`;

export const Image = styled.img`
  width: 100px;
  height: 75px;
  min-width: 100px;
  min-height: 75px;
  max-width: 100px;
  max-height: 75px;
  grid-area: image;
`;

export const Title = styled.div`
  font-weight: border;
  font-size: 1.2rem;
  grid-area: title;
`;

export const Description = styled.div`
  font-size: 1em;
  font-style: italic;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  max-height: 2em;
  max-width: 600px;
  grid-area: description;
`;

export const Price = styled.div`
  font-weight: border;
  font-size: 2rem;
  float: right;
  justify-self: center;
  align-self: center;
  grid-area: price;
`;
