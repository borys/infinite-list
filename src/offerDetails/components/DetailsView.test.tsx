import {} from './styled';

import { Offer, OfferStatus } from 'core/models';
import { ShallowWrapper, shallow } from 'enzyme';
import React from 'react';

import { DetailsView } from './DetailsView';
import { Description, Image, Price, Title } from './styled';

describe('DetailsListView', () => {
  let component: ShallowWrapper;
  const offer: Offer = {
    id: 1,
    title: 'title',
    description: 'description',
    img_url: 'image_url',
    price: 100,
    discount: 123,
    rating: 2,
    status: OfferStatus.Published,
    created_at: 'Tue Oct 15 2019 12:47:11 GMT+0200',
  };

  beforeEach(() => {
    component = shallow(<DetailsView goBack={() => true} offer={offer} />);
  });

  it('should render correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('should render title', () => {
    expect(component.contains(<Title>title</Title>)).toBe(true);
  });

  it('should render description', () => {
    expect(component.contains(<Description>description</Description>)).toBe(
      true
    );
  });

  it('should render price', () => {
    expect(
      component
        .find(Price)
        .first()
        .text()
    ).toContain('100');
  });

  it('should render image', () => {
    expect(
      component
        .find(Image)
        .first()
        .prop('src')
    ).toBe('image_url');
  });
});
