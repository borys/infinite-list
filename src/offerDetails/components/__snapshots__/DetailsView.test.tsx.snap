// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsListView should render correctly 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Go back
  </button>
  <styled.h1>
    title
  </styled.h1>
  <styled.img
    src="image_url"
  />
  <styled.span>
    description
  </styled.span>
  <styled.div>
    <styled.div>
      Cena:
      100
    </styled.div>
  </styled.div>
</div>
`;
